chasm/lib/activity/statemachine.go

```diff
@@ -280,11 +283,75 @@ var TransitionTerminated = chasm.NewTransition(
 	},
 )
 
+// TransitionCancelRequested affects a transition to CancelRequested status
+var TransitionCancelRequested = chasm.NewTransition(
+	[]activitypb.ActivityExecutionStatus{
+		activitypb.ACTIVITY_EXECUTION_STATUS_SCHEDULED,
+		activitypb.ACTIVITY_EXECUTION_STATUS_STARTED,
+		activitypb.ACTIVITY_EXECUTION_STATUS_CANCEL_REQUESTED, // Allow idempotent transition
+	},
+	activitypb.ACTIVITY_EXECUTION_STATUS_CANCEL_REQUESTED,
+	func(a *Activity, ctx chasm.MutableContext, req *activitypb.CancelActivityExecutionRequest) error {
```

Comment:
nit: take the frontend request as argument here so you don't have to convert it.